tests/test.py

```diff
@@ -475,8 +502,131 @@ def recur(j1, j2, path, unordered=False):
+        collected.
+
+        Optionally prints out the reference path to the object if it was not
+        garbage collected. This is enabled with the `WITH_REF_PATH` environment
+        variable.
+        """
+
+        gc.collect()
+
+        if msg is None:
+            msg = f"Object {ref} was not garbage collected."
+
+        obj = ref()
+
+        with self.subTest(part="garbage collection"):
+            self.assertTrue(obj is None, msg)
+
+        # Enable WITH_REF_PATH to see printout of why the given ref was not
+        # GC-ed.
+        if (
+            obj is not None
+            and os.environ.get(WITH_REF_PATH_VAR, None) is not None
+        ):
+            caller_globals = python_utils.caller_frame(offset=1).f_globals
+
+            with self.subTest(part="reference path"):
+                # Show the reference path to the given ref.
+                print(f"Reference path from globals to {ref}:")
+                path = find_path(id(caller_globals), id(obj))
+                self.assertIsNotNone(path, "Couldn't find reference path.")
+                print_referent_lens(path)
+
+    def tearDown(self):
+        """Check for running tasks and non-main threads after each test.
+
+        Raises:
+            AssertionError: If there are any running tasks running and the
+                environment variable `TEST_TASKS_CLEANUP` is set.
+
+            AssertionError: If there are any non-main threads running and the
+                environment variable `TEST_THREADS_CLEANUP` is set.
+        """
+
+        # GC here to make sure we don't have any references to tasks or threads
+        # that are keeping them alive.
+        gc.collect()
+
+        running_tasks = []
+        try:
+            loop = asyncio.get_event_loop()
+            for task in asyncio.all_tasks(loop):
+                running_tasks.append(task)
+        except Exception:
+            # No event loop running?
+            pass
+
+        if running_tasks:
+            if os.environ.get(TEST_TASKS_CLEANUP_VAR, None) is not None:
+                with self.subTest(part="running tasks"):
+                    raise AssertionError(
+                        f"Tasks still running: {running_tasks}"
+                    )
+            else:
+                print(f"Tasks still running: {running_tasks}")
+
+        non_main_threads = []
+        for thread in threading.enumerate():
+            if thread != threading.main_thread():
+                non_main_threads.append(thread)
+
+        if non_main_threads:
+            if os.environ.get(TEST_THREADS_CLEANUP_VAR, None) is not None:
+                with self.subTest(part="non-main threads"):
+                    raise AssertionError(
+                        f"Non-main threads still running: {non_main_threads}"
+                    )
+            else:
+                print(f"Non-main threads still running: {non_main_threads}")
+
+    @classmethod
+    def tearDownClass(cls):
```

Comment:
might want to call the super version too since it may do some good stuff? I have no idea.

Reply:
added at end